File: flutter/lib/services/socket_service.dart
import 'package:socket_io_client/socket_io_client.dart' as IO;
import '../config/app_config.dart';
import '../models/message.dart';

class SocketService {
  static final SocketService instance = SocketService._init();
  IO.Socket? _socket;
  bool _isConnected = false;

  // Callbacks
  Function(Map<String, dynamic>)? onMessageReceived;
  Function(String userId, bool isTyping)? onUserTyping;
  Function(String userId)? onUserOnline;
  Function(String userId)? onUserOffline;
  Function(String senderId, String messageType)? onNewMessageNotification;
  
  // Group chat callbacks
  Function(Message message)? onGroupMessageReceived;
  Function(Map<String, dynamic>)? onMemberJoinedGroup;
  Function(Map<String, dynamic>)? onMemberLeftGroup;

  SocketService._init();

  bool get isConnected => _isConnected;

  void connect(String token) {
    // If already connected, return
    if (_socket != null && _isConnected) {
      print('✓ Socket already connected');
      return;
    }

    // If socket exists but not connected, disconnect first
    if (_socket != null && !_isConnected) {
      print('🔄 Cleaning up old socket connection...');
      _socket!.dispose();
      _socket = null;
    }

    print('🔌 Connecting to socket: ${AppConfig.socketUrl}');
    
    _socket = IO.io(
      AppConfig.socketUrl,
      IO.OptionBuilder()
          .setTransports(['polling', 'websocket']) // Try polling first, then websocket
          .enableAutoConnect()
          .setAuth({'token': token})
          .build(),
    );

    _socket!.onConnect((_) {
      print('✓ Socket connected successfully');
      _isConnected = true;
    });

    _socket!.onDisconnect((_) {
      print('✗ Socket disconnected');
      _isConnected = false;
    });

    _socket!.onConnectError((error) {
      print('✗ Socket connection error: $error');
      _isConnected = false;
    });

    // Listen to message events
    _socket!.on('receive_message', (data) {
      print('📨 Message received: $data');
      if (onMessageReceived != null) {
        onMessageReceived!(Map<String, dynamic>.from(data));
      }
    });

    // Listen to typing events
    _socket!.on('user_typing', (data) {
      if (onUserTyping != null) {
        onUserTyping!(data['userId'], data['isTyping']);
      }
    });

    // Listen to online/offline events
    _socket!.on('user_online', (data) {
      if (onUserOnline != null) {
        onUserOnline!(data['userId']);
      }
    });

    _socket!.on('user_offline', (data) {
      if (onUserOffline != null) {
        onUserOffline!(data['userId']);
      }
    });

    // Listen to new message notifications
    _socket!.on('new_message_notification', (data) {
      if (onNewMessageNotification != null) {
        onNewMessageNotification!(data['senderId'], data['messageType']);
      }
    });

    _socket!.on('message_error', (data) {
      print('✗ Message error: $data');
    });

    // Group chat events
    _socket!.on('receive_group_message', (data) {
      print('📨 Group message received: $data');
      if (onGroupMessageReceived != null) {
        // Backend sends message data directly, not wrapped in 'message' field
        onGroupMessageReceived!(Message.fromJson(data));
      }
    });

    _socket!.on('member_joined', (data) {
      print('👤 Member joined group: ${data['username']}');
      if (onMemberJoinedGroup != null) {
        onMemberJoinedGroup!(Map<String, dynamic>.from(data));
      }
    });

    _socket!.on('member_left', (data) {
      print('👤 Member left group: ${data['username']}');
      if (onMemberLeftGroup != null) {
        onMemberLeftGroup!(Map<String, dynamic>.from(data));
      }
    });
  }

  void disconnect() {
    if (_socket != null) {
      _socket!.disconnect();
      _socket = null;
      _isConnected = false;
    }
  }

  // Join a conversation room
  void joinRoom(String receiverId) {
    if (_socket != null && _isConnected) {
      _socket!.emit('join_room', {'receiverId': receiverId});
      print('✓ Joined room with: $receiverId');
    }
  }

  // Send a message
  void sendMessage({
    required String receiverId,
    required String content,
    String messageType = 'text',
    bool isForwarded = false,
    String? originalSenderId,
    String? forwardedFrom,
    String? fileUrl,
    String? encryptedFileKey,
  }) {
    if (_socket != null && _isConnected) {
      final payload = {
        'receiverId': receiverId,
        'content': content,
        'messageType': messageType,
        'isForwarded': isForwarded,
        if (originalSenderId != null) 'originalSenderId': originalSenderId,
        if (forwardedFrom != null) 'forwardedFrom': forwardedFrom,
        if (fileUrl != null) 'fileUrl': fileUrl,
        if (encryptedFileKey != null) 'encryptedFileKey': encryptedFileKey,
      };
      _socket!.emit('send_message', payload);
      print('✉️ Message sent to: $receiverId');
    }
  }

  // Group chat methods
  void joinGroup(String roomId) {
    if (_socket != null && _isConnected) {
      _socket!.emit('join_group', {'roomId': roomId});
      print('✓ Joined group: $roomId');
    }
  }

  void leaveGroup(String roomId) {
    if (_socket != null && _isConnected) {
      _socket!.emit('leave_group', {'roomId': roomId});
      print('✓ Left group: $roomId');
    }
  }

  void sendGroupMessage({
    required String roomId,
    required String encryptedContent,
    required String iv,
    required String authTag,
    String messageType = 'text',
    String? fileUrl,
    String? encryptedFileKey,
  }) {
    if (_socket != null && _isConnected) {
      final payload = {
        'roomId': roomId,
        'content': encryptedContent,
        'iv': iv,
        'authTag': authTag,
        'messageType': messageType,
        if (fileUrl != null) 'fileUrl': fileUrl,
        if (encryptedFileKey != null) 'encryptedFileKey': encryptedFileKey,
      };

      _socket!.emit('send_group_message', payload);
      print('✓ Group message sent to room: $roomId');
    } else {
      print('✗ Socket not connected. Cannot send group message.');
    }
  }

  void sendGroupTyping(String roomId, bool isTyping) {
    if (_socket != null && _isConnected) {
      _socket!.emit('group_typing', {'roomId': roomId, 'isTyping': isTyping});
    }
  }

  // Set callback for group messages
  void onGroupMessage(Function(Message) callback) {
    onGroupMessageReceived = callback;
  }

  void onMemberJoined(Function(Map<String, dynamic>) callback) {
    onMemberJoinedGroup = callback;
  }

  void onMemberLeft(Function(Map<String, dynamic>) callback) {
    onMemberLeftGroup = callback;
  }

  // Send typing indicator
  void sendTyping(String receiverId, bool isTyping) {
    if (_socket != null && _isConnected) {
      _socket!.emit('typing', {'receiverId': receiverId, 'isTyping': isTyping});
    }
  }

  void dispose() {
    disconnect();
  }
}

File: backend/services/socketService.js
const jwt = require('jsonwebtoken');
const messageService = require('./messageService');
const User = require('../models/User');
const Room = require('../models/Room');
const RoomMember = require('../models/RoomMember');

class SocketService {
  constructor(io) {
    console.log('🚀 SocketService initializing...');
    this.io = io;
    this.users = new Map(); // userId -> socketId mapping
    this.setupSocketHandlers();
    console.log('✅ SocketService initialized successfully');
  }

  setupSocketHandlers() {
    console.log('📡 Setting up Socket.IO handlers...');
    
    this.io.use(async (socket, next) => {
      try {
        console.log('🔍 Socket connection attempt...');
        const token = socket.handshake.auth.token;
        
        if (!token) {
          console.log('❌ No token provided');
          return next(new Error('Authentication error'));
        }

        console.log('🔑 Verifying token...');
        const decoded = jwt.verify(token, process.env.JWT_SECRET);
        console.log('✅ Token verified for user:', decoded.userId);
        
        // Check if user is banned
        const user = await User.findById(decoded.userId);
        if (!user) {
          console.log('❌ User not found:', decoded.userId);
          return next(new Error('User not found'));
        }
        
        if (user.isBanned) {
          console.log('❌ User is banned:', decoded.userId);
          return next(new Error('Account is banned'));
        }
        
        socket.userId = decoded.userId;
        console.log('✅ Socket authenticated for user:', socket.userId);
        next();
      } catch (error) {
        console.log('❌ Socket auth error:', error.message);
        next(new Error('Authentication error'));
      }
    });

    console.log('🎧 Registering connection listener...');
    this.io.on('connection', (socket) => {
      console.log(`✓ User connected: ${socket.userId}`);
      
      // Store user connection
      this.users.set(socket.userId, socket.id);
      
      // Emit online status
      socket.broadcast.emit('user_online', { userId: socket.userId });

      // Join personal room
      socket.join(socket.userId);

      // Handle join room (conversation)
      socket.on('join_room', (data) => {
        this.handleJoinRoom(socket, data);
      });

      // Handle send message
      socket.on('send_message', async (data) => {
        await this.handleSendMessage(socket, data);
      });

      // Handle typing indicator
      socket.on('typing', (data) => {
        this.handleTyping(socket, data);
      });

      // ====== GROUP CHAT HANDLERS ======
      
      // Join group room
      socket.on('join_group', async (data) => {
        await this.handleJoinGroup(socket, data);
      });

      // Send group message
      socket.on('send_group_message', async (data) => {
        await this.handleSendGroupMessage(socket, data);
      });

      // Leave group room
      socket.on('leave_group', (data) => {
        this.handleLeaveGroup(socket, data);
      });

      // Group typing indicator
      socket.on('group_typing', (data) => {
        this.handleGroupTyping(socket, data);
      });

      // Handle disconnect
      socket.on('disconnect', () => {
        this.handleDisconnect(socket);
      });
    });
  }

  handleJoinRoom(socket, data) {
    const { receiverId } = data;
    const roomId = this.getRoomId(socket.userId, receiverId);
    socket.join(roomId);
    console.log(`✓ User ${socket.userId} joined room: ${roomId}`);
  }

  async handleSendMessage(socket, data) {
    try {
      const { 
        receiverId, 
        content, 
        messageType = 'text',
        isForwarded = false,
        originalSenderId,
        forwardedFrom,
        fileUrl,
        encryptedFileKey,
      } = data;

      // Save encrypted message to database (with forward fields)
      const message = await messageService.saveMessage({
        sender: socket.userId,
        receiver: receiverId,
        content, // Already encrypted by client
        messageType,
        isForwarded,
        originalSenderId,
        forwardedFrom,
        fileUrl,
        encryptedFileKey,
      });

      const roomId = this.getRoomId(socket.userId, receiverId);

      // Emit to room (both sender and receiver)
      this.io.to(roomId).emit('receive_message', {
        id: message._id,
        sender: socket.userId,
        receiver: receiverId,
        content: message.content,
        messageType: message.messageType,
        timestamp: message.timestamp,
        isForwarded: message.isForwarded,
        originalSenderId: message.originalSenderId,
        forwardedFrom: message.forwardedFrom,
        fileUrl: message.fileUrl,
        encryptedFileKey: message.encryptedFileKey,
      });

      // If receiver is online but not in room, send notification
      const receiverSocketId = this.users.get(receiverId);
      if (receiverSocketId) {
        this.io.to(receiverSocketId).emit('new_message_notification', {
          senderId: socket.userId,
          messageType: message.messageType,
          isForwarded: message.isForwarded,
        });
      }

      console.log(`✓ Message sent: ${socket.userId} -> ${receiverId} ${isForwarded ? '(forwarded)' : ''}`);
    } catch (error) {
      console.error('Send message error:', error);
      socket.emit('message_error', { error: 'Failed to send message' });
    }
  }


  // ============================================================
  // GROUP CHAT SOCKET HANDLERS
  // ============================================================

  /**
   * Handle user joining a group room
   * User must be verified as group member before joining socket room
   */
  async handleJoinGroup(socket, data) {
    try {
      const { roomId } = data;
      const userId = socket.userId;

      // Verify user is member of this group
      const room = await Room.findById(roomId);
      if (!room) {
        socket.emit('group_error', { error: 'Group not found' });
        return;
      }

      if (!room.isMember(userId)) {
        socket.emit('group_error', { error: 'You are not a member of this group' });
        return;
      }

      // Join socket room
      socket.join(roomId);
      
      // Notify other members
      socket.to(roomId).emit('member_joined', {
        roomId,
        userId,
        timestamp: Date.now()
      });

      console.log(`✓ User ${userId} joined group: ${roomId}`);
    } catch (error) {
      console.error('Join group error:', error);
      socket.emit('group_error', { error: 'Failed to join group' });
    }
  }

  /**
   * Handle sending message to group
   * Message is already encrypted with SessionKey by client
   */
  async handleSendGroupMessage(socket, data) {
    try {
      const { 
        roomId,
        content, // Encrypted with SessionKey
        messageType = 'text',
        iv, // Initialization vector for AES-GCM
        authTag, // Authentication tag for integrity
        fileUrl,
        encryptedFileKey,
        isForwarded = false,
        originalSenderId,
        forwardedFrom
      } = data;

      const userId = socket.userId;

      // Verify user is member
      const room = await Room.findById(roomId);
      if (!room || !room.isMember(userId)) {
        socket.emit('group_error', { error: 'Unauthorized' });
        return;
      }

      // Save message to database
      const Message = require('../models/Message');
      const message = new Message({
        sender: userId,
        receiver: null, // Group messages don't have single receiver
        roomId, // Group room ID
        content, // Encrypted content
        messageType,
        iv,
        authTag,
        fileUrl,
        encryptedFileKey,
        isForwarded,
        originalSenderId,
        forwardedFrom,
        timestamp: new Date()
      });

      await message.save();

      // Update room's last message time
      room.lastMessageAt = Date.now();
      await room.save();

      // Populate sender info
      const sender = await User.findById(userId).select('username email avatar');

      // Broadcast to all members in the room
      this.io.to(roomId).emit('receive_group_message', {
        id: message._id,
        roomId,
        sender: {
          id: sender._id,
          username: sender.username,
          email: sender.email,
          avatar: sender.avatar
        },
        content: message.content,
        messageType: message.messageType,
        iv: message.iv,
        authTag: message.authTag,
        fileUrl: message.fileUrl,
        encryptedFileKey: message.encryptedFileKey,
        isForwarded: message.isForwarded,
        originalSenderId: message.originalSenderId,
        forwardedFrom: message.forwardedFrom,
        timestamp: message.timestamp
      });

      console.log(`✓ Group message sent: ${userId} -> ${roomId} ${isForwarded ? '(forwarded)' : ''}`);
    } catch (error) {
      console.error('Send group message error:', error);
      socket.emit('group_error', { error: 'Failed to send message' });
    }
  }

  /**
   * Handle user leaving group room (socket disconnect from room)
   */
  handleLeaveGroup(socket, data) {
    try {
      const { roomId } = data;
      socket.leave(roomId);
      
      // Notify others
      socket.to(roomId).emit('member_left', {
        roomId,
        userId: socket.userId,
        timestamp: Date.now()
      });

      console.log(`✓ User ${socket.userId} left group: ${roomId}`);
    } catch (error) {
      console.error('Leave group error:', error);
    }
  }

  /**
   * Handle typing indicator for group chat
   */
  handleGroupTyping(socket, data) {
    try {
      const { roomId, isTyping } = data;
      const userId = socket.userId;

      // Broadcast to room (excluding sender)
      socket.to(roomId).emit('group_user_typing', {
        roomId,
        userId,
        isTyping
      });
    } catch (error) {
      console.error('Group typing error:', error);
    }
  }
  handleTyping(socket, data) {
    const { receiverId, isTyping } = data;
    const receiverSocketId = this.users.get(receiverId);
    
    if (receiverSocketId) {
      this.io.to(receiverSocketId).emit('user_typing', {
        userId: socket.userId,
        isTyping
      });
    }
  }

  handleDisconnect(socket) {
    console.log(`✗ User disconnected: ${socket.userId}`);
    this.users.delete(socket.userId);
    socket.broadcast.emit('user_offline', { userId: socket.userId });
  }

  // Generate consistent room ID for two users
  getRoomId(userId1, userId2) {
    return [userId1, userId2].sort().join('_');
  }
}

module.exports = SocketService;
